Faster nodal patch recovery

hugary1995 opened the pull request:
- Enable nodal patch recovery in parallel using RMs
- Refactor to avoid redundant material property evaluations

refs #15748
closes #12036

hugary1995 commented:
@lynnmunday I remember you are interested in this as well.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -845,7 +845,7 @@ FEProblemBase::initialSetup()
       TIME_SECTION("computingInitialStatefulProps", 3, "Computing Initial Material Values");
 
-      ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
+      ConstElemRange & elem_range = getEvaluableElementRange();
```
I grabbed this change from @fdkong 's branch which is fixing the issue mentioned in #12036. This is required for stateful properties on ghosted elements to be properly evaluated.

I think the additional computational cost is small. But if it is a concern, I can add a flag in FEProblem to only optionally evaluate nonlocal elements.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -845,7 +845,7 @@ FEProblemBase::initialSetup()
       TIME_SECTION("computingInitialStatefulProps", 3, "Computing Initial Material Values");
 
-      ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
+      const ConstElemRange & elem_range = getEvaluableElementRange();
```
I grabbed this change from @fdkong 's branch which is fixing the issue mentioned in #12036. This is required for stateful properties on ghosted elements to be properly evaluated.

I think the additional computational cost is small. But if it is a concern, I can add a flag in FEProblem to only optionally evaluate nonlocal elements.

hugary1995 replied:
Nope, I don't need this as long as I do a parallel pull in `NPRBase::finalize()`.

dschwen commented:
You did it!

hugary1995 commented:
I did a rough performance comparison by running the test with a finer mesh and more time steps:

The new nodal patch recovery: 
 - The node loop 0.149 seconds
 - The element loop 1.173 seconds

Using first order monomial:
 - The element loop 1.015 seconds

Using constant monomial:
 - The element loop 1.035 seconds

I think this new nodal patch recovery is now actually "usable".

hugary1995 commented:
Accuracy comparison (just by visualizing, I don't need to prove that NPR is super-convergent):

Nodal patch recovery
![stress_xx_npr](https://user-images.githubusercontent.com/31555204/131230687-d3bf6360-ad98-456d-81c1-9633a91862bf.png)

First order monomial
![stress_xx_nodal](https://user-images.githubusercontent.com/31555204/131230696-4a6cf7c1-1a2b-4b48-9b4f-1795c7e8c6e6.png)

Constant monomial
![stress_xx_elemental](https://user-images.githubusercontent.com/31555204/131230701-579c7b86-1f2b-4952-b2f2-3f4847acdf37.png)

hugary1995 commented:
I have to say that, if all you need is visualization, using the CellDataToPointData filter in paraview is probably the best choice:
![stress_xx_paraview](https://user-images.githubusercontent.com/31555204/131230793-de8fc223-0263-4a77-a450-662f6222edb4.png)

As you can see, the result is very similar to what was obtained with nodal patch recovery.

But if you need gradients of quantities defined at quadrature points during the simulation, nodal patch recovery is probably the way to go.

hugary1995 commented:
> You did it!

Yeah, this is implementing exactly what we discussed yesterday. The not-so-good part of this design is that we need to write a new `UserObject` to replace each `Auxkernel`, e.g. I wrote `NPRRankTwoAux` just to replace `RankTwoAux`. But I guess for tensor_mechanics, having several commonly UserObjects is probably enough.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/805996/) on 888963b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18721/site)

This comment will be updated on new commits.

dschwen commented:
`IndexableProperty` which I propose in https://github.com/idaholab/moose/pull/18661 should allow you to write one user object that can handle reals, vectors, and all rank tensors we have in one single object.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3884,7 +3884,7 @@ FEProblemBase::computeUserObjectsInternal(const ExecFlagType & type,
     // because some nodal user objects (NodalNormal related) depend on elemental user objects :-(
     ComputeUserObjectsThread cppt(*this, getNonlinearSystemBase(), query);
-    Threads::parallel_reduce(*_mesh.getActiveLocalElementRange(), cppt);
+    Threads::parallel_reduce(getEvaluableElementRange(), cppt);
```
This change is a no-go. This is going to mess with many calculations. As an example, you can imagine a elemental user object that computes a total domain volume will have the wrong result after this change because elements that are evaluable on multiple processes will multiply add their individual element volume. I can see the change to use evaluable elements for stateful material properties (`ComputeMaterialsObjectThread`), but that's the only loop I can see the change for.

hugary1995 commented on modules/tensor_mechanics/include/auxkernels/nodal_patch_recovery/NPRAux.h:
```diff
@@ -0,0 +1,28 @@
+
+class NPRBase;
+
+class NPRAux : public AuxKernel
```
> unless you think we're going to have 10 or more derived classes specifically from nodal patch recovery

I agree with you, but I'm afraid the answer to this is yes... If we have @dschwen 's `IndexableProperty` I could probably reduce the number of derived classes hence expand this class name.

dschwen replied:
We might have IndexableProperty before Alex' predicate makes it into libmesh (and MOOSE, through a submodule update)

lindsayad replied:
Probably since this libmesh update is a brutal one (we will be needing to patch around 50 tests in BISON)

hugary1995 replied:
I was discussing this with @bwspenc just today. It sounds like we are moving towards having a single material object that extracts all sorts of tensor_mechanics scalar quantities for output purposes (not sure which class is doing that though). So that is another option.

Either way, I think I don't need to add a whole bunch of userobjects to do nodal patch recovery, so I will remove the subdirectory and expand `NPR` in the class names.

hugary1995 replied:
> Probably since this libmesh update is a brutal one (we will be needing to patch around 50 tests in BISON)

I really hope this brutal update doesn't make the bison failure in #16816 worse...

dschwen commented:
Can you revert the change from `getActiveLocalElementRange` to `getEvaluableElementRange` and manually run a loop over all elements that are evaluable but not local?

lindsayad commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRRankTwoAux.C:
```diff
@@ -0,0 +1,44 @@
+registerMooseObject("TensorMechanicsApp", NPRRankTwoAux);
+
+InputParameters
+NPRRankTwoAux::validParams()
```
I don't get it... why is this called `Aux` when it's inheriting from a user object?

Ok you are concerned about having the user object loop over the evaluable element range ... but can you tell me why? Are you doing DG in your tensor mechanics computations? Right now it seems to me like you are going to be leveraging off-process material property information in your computation of *local* user object data, and this should now work (well after we get the libmesh links worked out). Are you planning on using *nonlocal* user object data somewhere? If you are, then we should think about what communication we want to do in `NPRBase::finalize`.

hugary1995 replied:
Yeah I'll think about a better name.

In `NPRAux.C`, for each node, I look at its connected elements. Therefore, the user object needs to have data from algebraically ghosted elements as well. That's my motivation of running user objects over evaluable elements.

Next, in that userobject, to compute the data on each element, I need to compute the requested material properties. That's my motivation of running materials over evaluable elements.

Now, on a second thought. I think I get away with all of these by doing the proper `finalize` in the user object, and keep the material loop and the user object loop still on local elements instead of evaluable elements. In `finalize`, I just need to loop over evaluable elements, and if the element is nonlocal, I can do the proper communication to grab the data from other processors.

Am I understanding it correctly now?

lindsayad replied:
> In `NPRAux.C`, for each node, I look at its connected elements. Therefore, the user object needs to have data from algebraically ghosted elements as well. That's my motivation of running user objects over evaluable elements.

We need to delineate consumption and production. I understand that you need the things your user object consumes (stateful material properties) to be computed on algebraically ghosted elements and hence why it makes sense to me to run the `ComputeMaterialsObjectThread` on evaluable elements ... this will ensure that the data your user object *consumes* are complete.

However, we must now turn to what your user object *produces*. Does *that* need to computed for algebraically ghosted elements? Are you going to have objects that need to get user object data from algebraically ghosted elements?

lindsayad replied:
My (perhaps very poor) conceptual understanding is that your nodal patch recovery user object is going to *produce* better *local* information by considering/consuming nonlocal information. But if you are doing something like DG, which will need to consume nonlocal data, then there I would understand the need to make sure your user object *produces* nonlocal data in addition to *consuming*  nonlocal data.

hugary1995 replied:
This is what my userobject produces:
```cpp
/// The element-level A matrix
std::vector<DenseMatrix<Real>> _Ae;
/// The element-level b vector
std::vector<DenseVector<Real>> _be;
virtual Real nodalPatchRecovery(const Point &, const std::vector<dof_id_type> &) const;
```
`_Ae` and `_be` are matrices and vectors for each element. The `NPRAux` will call `nodalPatchRecovery` with a point and a patch of elements. Some of the elements will be off-processor.

This is what my userobject consumes:
```
    be.add(computeValue(), p);
```
where `computeValue()` may need (stateful)material properties. See above, because `_be` should be available on some off-processor elements, _I thought_ what this user object consumes should also be available on those off-processor elements.

hugary1995 replied:
> are there things that are going to want the user object produced data for elements that are not local?

Yes, the `NPRAux` will want to call `NPRBase::nodalPatchRecovery` with elements that are not local.

lindsayad replied:
Ok yes I see. We'll definitely need the communication then, and I think pull will be the way to go

hugary1995 replied:
@dschwen Regarding the naming of this userobject: I completely agree with @lindsayad that calling a userobject `NPRRankTwoAux` is very confusing. 

I chose this name because I was thinking ahead about making actions. I think simply prepending `NPR` in front of `RankTwoAux` will make our lives easier when incorporating nodal patch recovery in TM action. But perhaps I should also add a suffix to tell people this is _NOT_ an auxkernel. Maybe `NPRRankTwoAuxUserObject`? What do you think?

dschwen replied:
`NPRRankTwoAuxUserObject` seems fine. Just a reminder though that we do have the `IndexableProperty` class available in the framework  now, which takes care of all teh indexing parameters and indexing and would make it trivial for this object to support all sorts of tensors (inclusding the upcoming Mandel notation tensors!).

dschwen replied:
The other option is to only provide this UO for Real material properties and have the user/action add a separate material that extracts the component (we do already have that). I don't like that option quite as much as my IndexableProperty solution (because it adds more clutter to the inputs if actions are not used), but in either case there should not be special objects to deal with Rank*Tensor components anymore.

hugary1995 replied:
Okay, good to know that `IndexableProperty` has made its way in! I'll try to use that.

BTW, what's a Mandel notation tensor?

hugary1995 commented:
> Can you revert the change from `getActiveLocalElementRange` to `getEvaluableElementRange` and manually run a loop over all elements that are evaluable but not local?

Yes, I think I get it now. See my reply to Alex above.

dschwen commented:
Alternatively you can do point to point communication in the user object!

dschwen commented:
I do this in https://github.com/idaholab/magpie/blob/devel/src/userobjects/RadialGreensConvolution.C

lindsayad commented:
@dschwen that code looks like a very good candidate for the push/pull family of functions in TIMPI. I think it could probably reduce the LOC by quite a bit. Yes @hugary1995 if you need nonlocal data `pull_parallel_vector_data` is going to be a good thing to look into. `pull` instead of `push` because an owning process does not know who needs its data. But the puller knows who owns the data it needs and so then can query and "pull" from the owner

dschwen commented:
> @dschwen that code looks like a very good candidate for the push/pull family of functions in TIMPI. I think it could probably reduce the LOC by quite a bit. 

Thanks for the pointer! (That code is a little over 3 years old - older than TIMPI :-D. But I should definitely check out the push/pull stuff!)

hugary1995 commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:
```diff
@@ -0,0 +1,166 @@
+}
+
+void
+NPRBase::finalize()
```
@lindsayad Do you see a type error here? I got the following compile error:
```
/home/gary/projects/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:165:62:   required from here
/home/gary/miniconda3/envs/moose/libmesh/include/timpi/standard_type.h:113:19: error: 'TIMPI::StandardType<T, Enable>::StandardType(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; Enable = void]' is private within this context
  113 |   StandardType<T> returnval(example);
      |                   ^~~~~~~~~
```
I see that the unspecialized StandardType constructor is private, but why do I hit that?

hugary1995 replied:
Nevermind, I figured it out. It turns out I need to use a datum that has a nullptr constructor.

lindsayad replied:
Can you paste in the whole backtrace? Conceptually I can imagine why there *might* be errors here. The "default" use of `push/pull_parallel_vector_data` is with "fixed" size types, and `DenseMatrix` is a dynamically sized type. However, I've tried to over time to make TIMPI dispatch to appropriate "dynamic" methods from "fixed-size" methods when it's required. At a minimum though, in order for this to work, we must have a `Packing` specialization of `DenseMatrix` and potentially `DenseVector` if you plan to be communicating that around in parallel. @roystgnr have you ever thought about doing `Packing` specializations for these? Off the top of my head I can't think of a barrier to entry here, but can never be too sure whether there are hidden gotchas

hugary1995 replied:
This is the error from civet [https://civet.inl.gov/job/806916/](https://civet.inl.gov/job/806916/).
I've worked around this using Eigen and passing around `std::vector<Real>`.
```
In file included from /opt/civet/build_2/moose/framework/build/header_symlinks/MooseUtils.h:25,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/InputParameters.h:13,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/DistributionInterface.h:12,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/ElementUserObject.h:13,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/build/header_symlinks/NPRBase.h:13,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:10,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/build/unity_src/userobjects_nodal_patch_recovery_Unity.C:2:
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h: In instantiation of 'TIMPI::StandardType<T> TIMPI::build_standard_type(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >]':
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/parallel_sync.h:383:34:   required from 'void TIMPI::push_parallel_vector_data(const TIMPI::Communicator&, MapToVectors&&, const ActionFunctor&) [with MapToVectors = std::multimap<unsigned int, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >, std::less<unsigned int>, std::allocator<std::pair<const unsigned int, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > > > > >&; ActionFunctor = TIMPI::pull_parallel_vector_data(const TIMPI::Communicator&, const MapToVectors&, GatherFunctor&, ActionFunctor&, const datum*) [with datum = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; MapToVectors = std::unordered_map<unsigned int, std::vector<long unsigned int> >; GatherFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>; ActionFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]::<lambda(TIMPI::processor_id_type, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]'
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/parallel_sync.h:486:29:   required from 'void TIMPI::pull_parallel_vector_data(const TIMPI::Communicator&, const MapToVectors&, GatherFunctor&, ActionFunctor&, const datum*) [with datum = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; MapToVectors = std::unordered_map<unsigned int, std::vector<long unsigned int> >; GatherFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>; ActionFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]'
/opt/civet/build_2/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:165:62:   required from here
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h:113:19: error: 'TIMPI::StandardType<T, Enable>::StandardType(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; Enable = void]' is private within this context
  113 |   StandardType<T> returnval(example);
      |                   ^~~~~~~~~
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h:85:3: note: declared private here
   85 |   StandardType(const T * example = nullptr);
      |   ^~~~~~~~~~~~
Compiling C++ (in opt mode) /opt/civet/build_2/moose/modules/navier_stokes/build/unity_src/fvbcs_Unity.C...
make: *** [/opt/civet/build_2/moose/framework/build.mk:122: /opt/civet/build_2/moose/modules/tensor_mechanics/build/unity_src/userobjects_nodal_patch_recovery_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
```

roystgnr replied:
The only possible catch I can see is the inheritance hierarchy we use with those classes to make the subvectors/submatrices work cleanly.  But I don't think that'll present any obstacle, and there's no other complications I can think of.

hugary1995 commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:
```diff
@@ -0,0 +1,173 @@
+
+  // Populate algebraically elements to querry
+  std::unordered_map<processor_id_type, std::vector<dof_id_type>> querry_ids;
+  const ConstElemRange evaluable_elem_range = _fe_problem.getEvaluableElementRange();
```
This is asking for evaluable elements based on the DofMap in nonlinear system. Ideally, this should be the intersection of {evaluable elements from nonlinear system} and {evaluable elements from auxiliary system}. We may need to wait for the next libmesh update where @lindsayad will add a predicate for that purpose.

lindsayad replied:
We will have to wait. I've written the code. Just need to do some testing and submit the PR

dschwen replied:
This: https://github.com/libMesh/libmesh/pull/3005 is the one, right? @hugary1995 will you be able to work on this anytime soon, or should we try and fix this up for you?

hugary1995 replied:
Yes, that's the one. I'm still waiting for a libmesh update #18831. Once it's merged I can work on this.

hugary1995 replied:
Oh wait... Maybe it's already merged.

hugary1995 replied:
Okay, I fixed it. This is ready for your review now.

hugary1995 commented:
There's an external app failing (probably due to the algebraic ghosting fix). Is it my responsibility to fix/regold the test in that app?

lindsayad commented:
> There's an external app failing (probably due to the algebraic ghosting fix). Is it my responsibility to fix/regold the test in that app?

Nah. I'll handle it

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -431,8 +432,18 @@ class FEProblemBase : public SubProblem, public Restartable
+   * getEvaluableElementRange() returns the element range that is evaluable based on both the
+   * nonlinear dofmap and the auxliary dofmap.
    */
   const ConstElemRange & getEvaluableElementRange();
+  const ConstElemRange & getNonlinearEvaluableElementRange();
+  const ConstElemRange & getAuxiliaryEvaluableElementRange();
```
So I renamed `getEvaluableElementRange` to `getNonlinearEvaluableElementRange` so that we can be honest about what this method actually does. `getEvaluableElementRange` now uses the multi evaluable predicate to construct an element range that is evaluable based on both the nonlinear dofmap and the aux dofmap.

Just to avoid diffs, I changed all old occurrences of `getEvaluableElementRange` to be `getNonlinearEvaluableElementRange`, although I'm not sure whether that's the original intent.

loganharbour replied:
I would prefer to not have `getEvaluableElementRange()` and `getAuxiliaryEvaluableElementRange()` if they are not needed.

hugary1995 replied:
Yeah, I'll remove `getAuxiliaryEvaluableElementRange()`. Then it's the same issue with `getEvaluableElementRange` -- it is in framework, but I'm testing it in tensor_mechanics. But this time it's hard for me to think of a test case in the framework...

lindsayad replied:
I think I would be in favor of lumping all our coverage together for a single `moose` coverage report but I see arguments both ways. It is hard when you have to very creatively invent framework level tests when there is a very natural module use case.

hugary1995 commented:
@recuero Do you have any idea about the pdass failure here: [https://civet.inl.gov/job/823622/](https://civet.inl.gov/job/823622/)? Could it be that the tolerance is not tight enough?

recuero commented:
It cuts the time step sporadically on CIVET, which prompts the failure. Tried to fix it by reducing the mesh size, but apparently didn't work. I'll add a fix to only check the final step. In the meantime, invalidating it should do the trick.

moosebuild commented:
Job [Test Heavy](https://civet.inl.gov/job/823622/) on 7151212 : invalidated by @hugary1995

Failing test is unrelated.

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/823625/) on 7151212 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9242cf</th>
      <th colspan=4>#18721 715121</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.51%</td>
      <td>81.46%</td>
      <td>-0.05%</td>
      <td>30.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>69416</td>
      <td>69424</td>
      <td>+8</td>
      <td>21</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15750</td>
      <td>15800</td>
      <td>+50</td>
      <td>49</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/framework.html)

## Modules coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9242cf</th>
      <th colspan=4>#18721 715121</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.04%</td>
      <td>87.06%</td>
      <td>+0.01%</td>
      <td>97.56%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91056</td>
      <td>91167</td>
      <td>+111</td>
      <td>120</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>13553</td>
      <td>13556</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/modules.html)

## Full coverage reports

<details><summary>Reports</summary>

- [framework](https://mooseframework.inl.gov/docs/PRs/18721/coverage/framework)
- [chemical_reactions](https://mooseframework.inl.gov/docs/PRs/18721/coverage/chemical_reactions)
- [combined](https://mooseframework.inl.gov/docs/PRs/18721/coverage/combined)
- [contact](https://mooseframework.inl.gov/docs/PRs/18721/coverage/contact)
- [external_petsc_solver](https://mooseframework.inl.gov/docs/PRs/18721/coverage/external_petsc_solver)
- [fluid_properties](https://mooseframework.inl.gov/docs/PRs/18721/coverage/fluid_properties)
- [fsi](https://mooseframework.inl.gov/docs/PRs/18721/coverage/fsi)
- [functional_expansion_tools](https://mooseframework.inl.gov/docs/PRs/18721/coverage/functional_expansion_tools)
- [geochemistry](https://mooseframework.inl.gov/docs/PRs/18721/coverage/geochemistry)
- [heat_conduction](https://mooseframework.inl.gov/docs/PRs/18721/coverage/heat_conduction)
- [level_set](https://mooseframework.inl.gov/docs/PRs/18721/coverage/level_set)
- [misc](https://mooseframework.inl.gov/docs/PRs/18721/coverage/misc)
- [navier_stokes](https://mooseframework.inl.gov/docs/PRs/18721/coverage/navier_stokes)
- [peridynamics](https://mooseframework.inl.gov/docs/PRs/18721/coverage/peridynamics)
- [phase_field](https://mooseframework.inl.gov/docs/PRs/18721/coverage/phase_field)
- [porous_flow](https://mooseframework.inl.gov/docs/PRs/18721/coverage/porous_flow)
- [ray_tracing](https://mooseframework.inl.gov/docs/PRs/18721/coverage/ray_tracing)
- [rdg](https://mooseframework.inl.gov/docs/PRs/18721/coverage/rdg)
- [richards](https://mooseframework.inl.gov/docs/PRs/18721/coverage/richards)
- [stochastic_tools](https://mooseframework.inl.gov/docs/PRs/18721/coverage/stochastic_tools)
- [tensor_mechanics](https://mooseframework.inl.gov/docs/PRs/18721/coverage/tensor_mechanics)
- [xfem](https://mooseframework.inl.gov/docs/PRs/18721/coverage/xfem)

</details>


This comment will be updated on new commits.

loganharbour commented:
Is https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/framework.html#framework/src/utils/MathUtils.C committed on accident? not used anywhere

roystgnr commented:
There are unit tests for it in `unit/src/MathUtilsTest.C` at least.  I can't find any other internal use cases either, but the functions there make sense as things to offer to app codes.

roystgnr commented:
Well, there are unit tests for much of it at least.  I can't find anything that's testing invocations of (or [the spelling of](https://en.wikipedia.org/wiki/Oliver_Heaviside)) the `heavyside()` function there.

loganharbour commented:
Yeah - exactly. I don't disagree that they're not useful... except they're not used or tested.

roystgnr commented:
Whoa, I should have clicked the link before commenting.  Those specific new functions have to be an accidental commit, right?  They're not used in the source file, and without a corresponding MathUtils.h declaration they're not usable anywhere else either!

hugary1995 commented:
`MathUtils::multiIndex` is used here: [https://github.com/idaholab/moose/blob/7151212dc919cd69b634d6436668c9472b4782a2/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C#L47](https://github.com/idaholab/moose/blob/7151212dc919cd69b634d6436668c9472b4782a2/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C#L47)

I've no idea why gcov didn't record it.

hugary1995 commented:
> Whoa, I should have clicked the link before commenting. Those specific new functions have to be an accidental commit, right? They're not used in the source file, and without a corresponding MathUtils.h declaration they're not usable anywhere else either!

They have declarations in `MathUtils.h`

loganharbour commented:
You must test it within the framework. Framework coverage only comes from tests that are ran from within the framework. Please add/move a unit test the framework unit tests.

roystgnr commented:
Ah, I see the declarations now.  Not sure if I was looking at the wrong branch or how else I missed them before; sorry.

And yeah: NPRBase uses multiIndex, multiIndex uses nChooseK, and the new patch_recovery.i should be reaching that.  The problem is just that it's only in a modules test and we're only evaluating coverage from test/ ?

loganharbour commented:
>  The problem is just that it's only in a modules test and we're only evaluating coverage from test/

Almost. Framework tests come from `test/` and `unit/`. The intention of course is that we should test framework code from within framework tests, and not from dependent applications.

hugary1995 commented:
Okay, I see. Thanks to both of you. Before I make any changes -- do you think these functions are general enough to be moved into the framework? If not, I can move it to a utility class in tensor_mechanics.

loganharbour commented:
I think they're both general and appropriate given proper testing 👍🏼

loganharbour commented:
@dschwen, @lindsayad, or @bwspenc someone want to review?

lindsayad reviewed (dismissed):
Looks good to me. I like to see the use of the new `multi_evaluable` predicate and more `push_pull`. Is there a patch ready for the failing apps?

hugary1995 commented:
> Looks good to me. I like to see the use of the new `multi_evaluable` predicate and more `push_pull`. Is there a patch ready for the failing apps?

I don't have a patch ready. @lindsayad said he'll handle it.

The diff in Numbat seems to be related to the periodic BC changes you made in this PR. If I were to create a patch, all I can do is probably regold the test, as I am not sure if the diff is improving/correcting the solution or not.

lindsayad commented:
Oh yea I did, didn't I?

On Wed, Sep 29, 2021 at 6:29 PM Gary (Tianchen) Hu ***@***.***>
wrote:

> Looks good to me. I like to see the use of the new multi_evaluable
> predicate and more push_pull. Is there a patch ready for the failing apps?
>
> I don't have a patch ready. @lindsayad <https://github.com/lindsayad>
> said he'll handle it.
>
> The diff in Numbat seems to be related to the periodic BC changes you made
> in this PR. If I were to create a patch, all I can do is probably regold
> the test, as I am not sure if the diff is improving/correcting the solution
> or not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/18721#issuecomment-930676043>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4G6RW4NUS6DJ3IXZWDUEO4PXANCNFSM5C7NPIGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

lindsayad commented:
Ok I have the patch ready to go. I would prefer to have someone a little more in the physics area approve this as well. @dschwen could you add your stamp of approval please?

dschwen commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:
```diff
@@ -0,0 +1,173 @@
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
+  {
+    polynomial = 1.0;
+    for (unsigned int c = 0; c < _multi_index[0].size(); c++)
```
`_multi_index[r].size()` better be `_mesh.dimension()` for every `r`, right? I'd suggest adding a `_dim` member, using it here, and adding a `mooseAssert(_multi_index[r].size() == _dim, "Wrong moose index element size");` for posterity.

dschwen replied:
... and make it 
```suggestion
    {
      mooseAssert(index.size() == _dim, "Wrong moose index element size"); 
      for (unsigned int c = 0; c < _dim; c++)
```

dschwen commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:
```diff
@@ -0,0 +1,173 @@
+{
+  RealEigenVector p(_q);
+  Real polynomial;
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
```
Or even better use 
```suggestion
  for (const auto & index: _multi_index)
```
here

dschwen commented on modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:
```diff
@@ -0,0 +1,173 @@
+    for (unsigned int c = 0; c < _multi_index[0].size(); c++)
+      for (unsigned int p = 0; p < _multi_index[r][c]; p++)
+        polynomial *= q_point(c);
+    p(r) = polynomial;
```
nevermind you need `r` here

hugary1995 commented:
@dschwen I switched to `IndexableProperty` in `NodalPatchRecoveryMaterialProperty.C` but I am getting:
```
The non-AD material property 'stress' does not exist
```
Do you know what might be wrong?

moosebuild commented:
All [jobs](https://civet.inl.gov/event/76295/) on 30f0130 : canceled by @hugary1995

Don&#39;&#39;t run any test before I fix the bug.

dschwen commented:
Weird, looks good to me. But I'm out camping and can only take a cursory look. Will check again on Monday.

hugary1995 commented:
Hopefully this will save you some detective work:

I set a breakpoint here:
```
Thread 1 "tensor_mechanic" hit Breakpoint 1, IndexableProperty<NodalPatchRecoveryBase, false>::getPropertyHelper<RankTwoTensorTempl<double> > (this=0x555555ae89c0, pointer=@0x555555ae89d8: 0x0, components=2)
    at /home/gary/projects/moose/framework/build/header_symlinks/IndexableProperty.h:138
138	  if (_host->template hasGenericMaterialProperty<P, is_ad>(_property_param))
```
and
```cpp
_host->template hasGenericMaterialProperty<P, is_ad>(_property_param)
```
returns false. I stepped in to check what's going on: `MaterialPropertyInterface` actually successfully  deduced the property name from `_property_param = "property"` as `"stress"`, but when it does look up in `MaterialPropertyStorage`, the size of `MaterialPropertyStorage::_prop_ids` is zero.

dschwen commented:
Can you try getting the same material property directly in you uo? Just to
check if the getMaterialProperty call works?

On Sun, Oct 3, 2021, 7:12 AM Gary (Tianchen) Hu ***@***.***>
wrote:

> Hopefully this will save you some detective work:
>
> I set a breakpoint here:
>
> Thread 1 "tensor_mechanic" hit Breakpoint 1, IndexableProperty<NodalPatchRecoveryBase, false>::getPropertyHelper<RankTwoTensorTempl<double> > (this=0x555555ae89c0, ***@***.***: 0x0, components=2)
>     at /home/gary/projects/moose/framework/build/header_symlinks/IndexableProperty.h:138
> 138	  if (_host->template hasGenericMaterialProperty<P, is_ad>(_property_param))
>
> and
>
> _host->template hasGenericMaterialProperty<P, is_ad>(_property_param)
>
> returns false. I stepped in to check what's going on:
> MaterialPropertyInterface actually successfully deduced the property name
> from _property_param = "property" as "stress", but when it does look up
> in MaterialPropertyStorage, the size of MaterialPropertyStorage::_prop_ids
> is zero.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/18721#issuecomment-932948640>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPTJW44TKS7SL6XF6ZTUFBJE5ANCNFSM5C7NPIGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

hugary1995 commented:
`_prop(getMaterialProperty<RankTwoTensor>("property"))` works fine.

dschwen commented:
Ugh, crap. Of course. User objects are constructed before materials so hasMaterialProperty will always return false. :-(

hugary1995 commented:
So, without this check, everything else in IndexableProperty still works in a UO because the right pointer will still be properly initialized after construction, right? Perhaps we can add a flag to skip the check at construction time, and provide a public method to manually perform the check.

dschwen commented:
No, without that check all material properties will be requested and then
the dependency resolution will throw an error. We need a whine new method
of getting material properties that returns a reference to a material
property pointer that stays zero if the property does not exist.

On Sun, Oct 3, 2021, 10:28 PM Gary (Tianchen) Hu ***@***.***>
wrote:

> So, without this check, everything else in IndexableProperty still works
> in a UO because the right pointer will still be properly initialized after
> construction, right? Perhaps we can add a flag to skip the check at
> construction time, and provide a public method to manually perform the
> check.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/18721#issuecomment-933133962>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPXQVX5QP3RI5HMCWVLUFEUQFANCNFSM5C7NPIGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

hugary1995 commented:
Okay that makes sense. If you think you could work on that alternative material getter method, this PR can wait. Otherwise I'll switch back to the original implementation without the IndexableProperpty for now.

dschwen commented:
My suggestion is to switch to a Real only implementation for now and use
the RankTwoCartesianComponent material to extract the component.

On Mon, Oct 4, 2021, 7:51 AM Gary (Tianchen) Hu ***@***.***>
wrote:

> Okay that makes sense. If you think you could work on that alternative
> material getter method, this PR can wait. Otherwise I'll switch back to the
> original implementation without the IndexableProperpty for now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/18721#issuecomment-933507195>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPXQ3CTHQVSPTPWJEMDUFGWNJANCNFSM5C7NPIGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

hugary1995 commented:
> No, without that check all material properties will be requested and then the dependency resolution will throw an error. We need a whine new method of getting material properties that returns a reference to a material property pointer that stays zero if the property does not exist.

I've heard about some ongoing work about DAG several times. I would imagine this issue could be easily solved if construction order is scheduled by a topological sorting. Is that what the DAG work in moose is about? If so, how far are we from being able to use DAG? One year? Two?

lindsayad commented:
@rwcarlsen has done the research so I’ll let him comment but I wouldn’t count on it anytime soon

dschwen commented:
https://github.com/idaholab/moose/issues/18999 viel help

dschwen commented:
@lindsayad did you want to fix that _numbat_ fail? Who's app is that anyways?

lindsayad commented:
It's @cpgr's app I think. It just requires a simple regold. This has been up here for so long I forget why, but I remember thinking it was a good reason

lindsayad commented:
@hugary1995 can you fix the merge conflict? Sorry that we have not gotten this in for you yet. @cpgr would you be willing to create a patch for Numbat based on this PR? I am making Numbat allowed to fail on the external app recipe anyway because it doesn't have a MOOSE submodule so we can't give it the standard next branch treatment.

Once the merge conflict is fixed, I will merge this

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit cfbbbde0d0e51ce94f75615784cd9a6867159aa8
+Subproject commit 8d3358bb335f9d3c9aa6476c5ea1bd0a999d6f36
```
**Caution!** This contains a submodule update

hugary1995 replied:
This was by mistake. I reverted it.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/892755/) on 25fb482 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

cpgr commented:
> @hugary1995 can you fix the merge conflict? Sorry that we have not gotten this in for you yet. @cpgr would you be willing to create a patch for Numbat based on this PR? I am making Numbat allowed to fail on the external app recipe anyway because it doesn't have a MOOSE submodule so we can't give it the standard next branch treatment.
> 
> Once the merge conflict is fixed, I will merge this

I just skipped this test so that your testing will pass now. I'll fix it up when this PR is in.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/892781/) on 25fb482 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18721 25fb48</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>81.95%</td>
      <td>-0.00%</td>
      <td>84.38%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72274</td>
      <td>72314</td>
      <td>+40</td>
      <td>54</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15915</td>
      <td>15926</td>
      <td>+11</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage/framework)


## Modules coverage

### Geochemistry

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18721 25fb48</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.59%</td>
      <td>96.59%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4822</td>
      <td>4822</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>170</td>
      <td>170</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/geochemistry.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage/geochemistry)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18721 25fb48</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.80%</td>
      <td>89.80%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11871</td>
      <td>11871</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1349</td>
      <td>1349</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/phase_field.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage/phase_field)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18721 25fb48</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.09%</td>
      <td>96.09%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10492</td>
      <td>10492</td>
      <td>-</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>427</td>
      <td>427</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/porous_flow.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage/porous_flow)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>44d8c6</th>
      <th colspan=4>#18721 25fb48</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.83%</td>
      <td>88.86%</td>
      <td>+0.03%</td>
      <td>97.30%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>24793</td>
      <td>24901</td>
      <td>+108</td>
      <td>108</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3118</td>
      <td>3121</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18721/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18721/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 84.38% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit cfbbbde0d0e51ce94f75615784cd9a6867159aa8
+Subproject commit 8d3358bb335f9d3c9aa6476c5ea1bd0a999d6f36
```
**Caution!** This contains a submodule update

hugary1995 replied:
This was by mistake. I reverted it.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/899240/) on 6d700bb wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

hugary1995 committed:
Faster nodal patch recovery

- Enable nodal patch recovery in parallel using RMs
- Refactor to avoid redundant material property evaluations
- Use the modern push/pull during finalize() of NodalPatchRecoveryBase to
communicate precomputed elemental least squares coefficients and values.

refs idaholab#15748
closes idaholab#12036

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

hugary1995 commented:
@dschwen with #19000 can I use IndexableProperty now?

dschwen commented:
Yes!

hugary1995 committed:
Attempt to use IndexableProperty

hugary1995 commented:
A warm greeting when you wake up tomorrow: The IndexableProperty is giving me the same error
```
test:nodal_patch_recovery.nodal_patch_recovery: *** ERROR ***
test:nodal_patch_recovery.nodal_patch_recovery: The following error occurred in the object "stress_xx_patch", of type "NodalPatchRecoveryMaterialProperty".
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery: The non-AD material property 'stress' does not exist
```

If I do not check the validity of the indexable property at `initialSetup`, then it solves the first step then I get the following error:
```
test:nodal_patch_recovery.nodal_patch_recovery: Time Step 1, time = 0.05, dt = 0.05
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery: Performing automatic scaling calculation
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery:  0 Nonlinear |R| = 1.118034e-03
test:nodal_patch_recovery.nodal_patch_recovery:  1 Nonlinear |R| = 9.428704e-08
test:nodal_patch_recovery.nodal_patch_recovery:  2 Nonlinear |R| = 2.247769e-14
test:nodal_patch_recovery.nodal_patch_recovery:  Solve Converged!
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery: *** ERROR ***
test:nodal_patch_recovery.nodal_patch_recovery: The following error occurred in the object "stress_xx_patch", of type "NodalPatchRecoveryMaterialProperty".
test:nodal_patch_recovery.nodal_patch_recovery: 
test:nodal_patch_recovery.nodal_patch_recovery: internal error in IndexableProperty
```

dschwen commented:
I'll check out you branch tomorrow.

On Thu, Dec 2, 2021, 9:52 PM Gary (Tianchen) Hu ***@***.***>
wrote:

> A warm greeting when you wake up tomorrow: The IndexableProperty is giving
> me the same error
>
> test:nodal_patch_recovery.nodal_patch_recovery: *** ERROR ***
> test:nodal_patch_recovery.nodal_patch_recovery: The following error occurred in the object "stress_xx_patch", of type "NodalPatchRecoveryMaterialProperty".
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery: The non-AD material property 'stress' does not exist
>
> If I do not check the validity of the indexable property at initialSetup,
> then it solves the first step then I get the following error:
>
> test:nodal_patch_recovery.nodal_patch_recovery: Time Step 1, time = 0.05, dt = 0.05
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery: Performing automatic scaling calculation
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery:  0 Nonlinear |R| = 1.118034e-03
> test:nodal_patch_recovery.nodal_patch_recovery:  1 Nonlinear |R| = 9.428704e-08
> test:nodal_patch_recovery.nodal_patch_recovery:  2 Nonlinear |R| = 2.247769e-14
> test:nodal_patch_recovery.nodal_patch_recovery:  Solve Converged!
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery: *** ERROR ***
> test:nodal_patch_recovery.nodal_patch_recovery: The following error occurred in the object "stress_xx_patch", of type "NodalPatchRecoveryMaterialProperty".
> test:nodal_patch_recovery.nodal_patch_recovery:
> test:nodal_patch_recovery.nodal_patch_recovery: internal error in IndexableProperty
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/18721#issuecomment-985215377>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRMPVU7L6PU3DCBWPH2WTUPBEJZANCNFSM5C7NPIGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

dschwen commented:
D'Oh, ok, I get it. That's a design flaw that I need to think about a bit.

I have `getOptionalMaterialProperty` methods in `Material.h` and `MaterialPropertyInterface.h`. The methods in Material do the late binding, the methods in MaterialPropertyInterface are for objects that run later when all materials exist. 

This of course does not take care of objects that run entirely _before_ the first material is even constructed.

@loganharbour that's why I actually wanted to store the proxies in FEProblem. We'll have to come up with a better design that takes care of that 3rd phase of objects that are constructed even earlier than materials. 

I should have thought of the UserObject case! My tests for that PR only use Material and AuxKernel. Dang!

dschwen commented:
#19546 will fix this

hugary1995 committed:
Minor doco fix

hugary1995 commented:
Writing this PR has been fun. New capabilities like the parallel push/pull interfaces and IndexableProperty are making my life so much easier as a downstream developer.

dschwen commented on framework/include/utils/MathUtils.h:
```diff
@@ -316,4 +316,26 @@ mooseSetToZero(std::vector<Real> & vec)
+ * @param order generate the multi-index up to certain order
+ * @return a data structure holding entries representing the complete multi index
+ */
+std::vector<std::vector<unsigned int>> multiIndex(unsigned int dim, unsigned int order);
```
Order here is a polynomial order, right? When I first read `multiIndex` I thought it was an unrolled way to iterate over rank N tensor entries.

hugary1995 replied:
It could be a polynomial order. This method gives all multi indices up to order `order`. Say we have a multi index `a`, then this method gives all multi indices with `|a| <= order`. Perhaps there's a better name, but I think the description is clear enough.

dschwen reviewed (dismissed):
Looks fine to me.

hugary1995 committed:
Fix typos

bwspenc commented:
I'm really excited to have this finally merged in!

One minor follow-on -- this broke the documentation builds for the apps that do sqa checks because that test points to `nodal_patch_recovery.md`, which is not in tensor_mechanics. We've had to go back and specifically add this file in to the config files for the apps that depend on tensor_mechanics. We really should fix this by making that test point to something that's already included by the apps.

bwspenc commented:
For example, see this issue: https://github.com/idaholab/blackbear/issues/285

hugary1995 commented:
So in principal the tests in a module should only point to docs in that module (or dependent modules). Otherwise, apps that depend on that module will have to explicitly include those docs in the config. Am I understanding it correctly?

I guess we can either add ` - modules/doc/content` to the tensor mechanics doc config file, or move `modules/doc/content/help/finite_element_concepts/nodal_patch_recovery.md` to somewhere in the `framework/doc/content`
